File: cdisc_transpiler/dataset_xml.py
"""Dataset-XML 1.0 generation module.

This module generates Dataset-XML 1.0 compliant files following the
CDISC Dataset-XML 1.0 specification for SDTM data exchange.

Dataset-XML 1.0 Key Features:
- Uses ODM 1.3.2 as the base with Dataset-XML extensions
- data:DatasetXMLVersion attribute on ODM root (required)
- ClinicalData container for subject-level data
- ReferenceData container for trial design data
- ItemGroupData elements with data:ItemGroupDataSeq attribute
- ItemData elements with ItemOID and Value attributes
- Supports streaming/chunked output for large datasets

Namespace definitions:
- ODM: http://www.cdisc.org/ns/odm/v1.3
- data: http://www.cdisc.org/ns/Dataset-XML/v1.0

Reference:
- CDISC Dataset-XML 1.0 specification
- docs/Dataset-XML_1-0/Example/ for examples
"""

from __future__ import annotations

import xml.etree.ElementTree as ET
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import TYPE_CHECKING, Iterator

if TYPE_CHECKING:
    import pandas as pd

    from cdisc_transpiler.mapping import MappingConfig

from cdisc_transpiler.domains import SDTMDomain, get_domain

# XML Namespaces for Dataset-XML 1.0
ODM_NS = "http://www.cdisc.org/ns/odm/v1.3"
DATA_NS = "http://www.cdisc.org/ns/Dataset-XML/v1.0"
XLINK_NS = "http://www.w3.org/1999/xlink"

# Dataset-XML version
DATASET_XML_VERSION = "1.0.0"

# Define-XML version for PriorFileOID reference
DEFINE_XML_VERSION = "2.1.0"
DEFAULT_SDTM_VERSION = "3.4"

# Shared variables that use IT.{VARIABLE} OID pattern (no domain prefix)
# RDOMAIN follows the shared pattern per CDISC examples (see SUPP-- datasets).
SHARED_VARIABLE_OIDS = {"STUDYID", "USUBJID", "RDOMAIN"}


class DatasetXMLError(Exception):
    """Error during Dataset-XML generation."""


@dataclass
class DatasetXMLConfig:
    """Configuration for Dataset-XML generation."""

    study_oid: str
    metadata_version_oid: str
    originator: str = "CDISC-Transpiler"
    source_system: str = "CDISC-Transpiler"
    source_system_version: str = "1.0"
    file_type: str = "Snapshot"
    granularity: str = "All"  # "All", "Metadata", "AdminData", "ReferenceData", "AllClinicalData", "SingleSite"


def _tag(namespace: str, name: str) -> str:
    """Return a fully qualified XML tag."""
    return f"{{{namespace}}}{name}"


def _attr(namespace: str, name: str) -> str:
    """Return a fully qualified XML attribute."""
    return f"{{{namespace}}}{name}"


def _register_namespaces() -> None:
    """Register XML namespaces for pretty output."""
    ET.register_namespace("", ODM_NS)
    ET.register_namespace("xlink", XLINK_NS)
    ET.register_namespace("data", DATA_NS)


def build_dataset_xml_tree(
    data: pd.DataFrame,
    domain_code: str,
    config: MappingConfig,
    *,
    metadata_version_oid: str | None = None,
    is_reference_data: bool = False,
) -> ET.Element:
    """Build a Dataset-XML 1.0 document tree for a single domain.

    Args:
        data: The pandas DataFrame containing the domain data.
        domain_code: The SDTM domain code (e.g., "DM", "AE").
        config: The mapping configuration containing study metadata.
        metadata_version_oid: The MetaDataVersionOID to reference Define-XML.
        is_reference_data: Whether this is reference data (trial design).

    Returns:
        The root Element of the Dataset-XML 1.0 document.

    Per CDISC Dataset-XML 1.0 standard:
    - FileOID references the ItemGroupDef in the Define-XML
    - PriorFileOID references the Define-XML document
    - StudyOID matches the Study/@OID in Define-XML
    - MetaDataVersionOID matches the MetaDataVersion/@OID in Define-XML
    """
    _register_namespaces()

    domain = get_domain(domain_code)
    study_id = (config.study_id or "STUDY").strip() or "STUDY"
    study_oid = f"STDY.{study_id}"
    dataset_name = domain.resolved_dataset_name()
    timestamp = datetime.now(UTC).isoformat(timespec="seconds")

    # Determine MetaDataVersionOID - must match Define-XML MetaDataVersion/@OID
    mdv_oid = metadata_version_oid or f"MDV.{study_oid}.SDTMIG.{DEFAULT_SDTM_VERSION}"

    # Define-XML FileOID for PriorFileOID reference
    define_file_oid = f"{study_oid}.Define-XML_{DEFINE_XML_VERSION}"

    # Build ODM root element per Dataset-XML 1.0 spec
    root = ET.Element(
        _tag(ODM_NS, "ODM"),
        attrib={
            "FileType": "Snapshot",
            # FileOID identifies this dataset file and references ItemGroupDef
            "FileOID": f"{define_file_oid}(IG.{dataset_name})",
            # PriorFileOID references the Define-XML document
            "PriorFileOID": define_file_oid,
            "ODMVersion": "1.3.2",
            "CreationDateTime": timestamp,
            "Originator": "CDISC-Transpiler",
        },
    )
    # Explicitly declare xlink namespace for completeness (matches CDISC examples)
    root.set("xmlns:xlink", XLINK_NS)
    # data:DatasetXMLVersion is REQUIRED in Dataset-XML 1.0
    root.set(_attr(DATA_NS, "DatasetXMLVersion"), DATASET_XML_VERSION)

    # Choose container based on data type
    if is_reference_data:
        # ReferenceData for trial design domains (TA, TE, TV, TI, TS, SE, SV)
        container = ET.SubElement(
            root,
            _tag(ODM_NS, "ReferenceData"),
            attrib={
                # StudyOID must match Study/@OID in Define-XML (without STDY. prefix per standard examples)
                "StudyOID": study_oid,
                "MetaDataVersionOID": mdv_oid,
            },
        )
    else:
        # ClinicalData for subject-level domains
        container = ET.SubElement(
            root,
            _tag(ODM_NS, "ClinicalData"),
            attrib={
                # StudyOID must match Study/@OID in Define-XML (without STDY. prefix per standard examples)
                "StudyOID": study_oid,
                "MetaDataVersionOID": mdv_oid,
            },
        )

    # Add ItemGroupData elements
    _append_item_group_data(container, data, domain, dataset_name)

    return root


def _append_item_group_data(
    parent: ET.Element,
    data: pd.DataFrame,
    domain: SDTMDomain,
    dataset_name: str,
) -> None:
    """Append ItemGroupData elements for each row in the DataFrame.

    Per Dataset-XML 1.0:
    - ItemGroupData has data:ItemGroupDataSeq attribute (sequence number)
    - ItemGroupData has ItemGroupOID pointing to ItemGroupDef
    - ItemData elements contain ItemOID and Value

    ItemOID conventions per CDISC standard:
    - Shared variables (STUDYID, USUBJID): IT.{VARIABLE}
    - Domain-specific variables: IT.{DOMAIN}.{VARIABLE}
    """
    for seq, (_, row) in enumerate(data.iterrows(), start=1):
        # Create ItemGroupData with sequence number
        item_group_data = ET.SubElement(
            parent,
            _tag(ODM_NS, "ItemGroupData"),
            attrib={
                "ItemGroupOID": f"IG.{dataset_name}",
            },
        )
        # data:ItemGroupDataSeq is the row sequence number
        item_group_data.set(_attr(DATA_NS, "ItemGroupDataSeq"), str(seq))

        # Add ItemData for each column
        for col_name in data.columns:
            value = row[col_name]

            # Skip null/NaN values
            if _is_null(value):
                continue

            # Format value appropriately
            formatted_value = _format_value(value, col_name)

            # Generate ItemOID following CDISC standard conventions
            item_oid = _generate_item_oid(col_name, domain.code)

            ET.SubElement(
                item_group_data,
                _tag(ODM_NS, "ItemData"),
                attrib={
                    "ItemOID": item_oid,
                    "Value": formatted_value,
                },
            )


def _generate_item_oid(variable_name: str, domain_code: str) -> str:
    """Generate ItemOID following CDISC standard conventions.

    Per CDISC Dataset-XML 1.0 standard:
    - Shared variables (STUDYID, USUBJID) use IT.{VARIABLE} without domain prefix
    - Domain-specific variables use IT.{DOMAIN}.{VARIABLE}
    """
    name = variable_name.upper()
    if name in SHARED_VARIABLE_OIDS:
        return f"IT.{name}"
    return f"IT.{domain_code.upper()}.{variable_name}"


def _is_null(value: object) -> bool:
    """Check if a value is null/NaN/empty."""
    import pandas as pd

    if value is None:
        return True
    if isinstance(value, float) and pd.isna(value):
        return True
    if isinstance(value, str) and value.strip() == "":
        return True
    return False


def _format_value(
    value: object,
    column_name: str,
) -> str:
    """Format a value for Dataset-XML output.

    Handles proper formatting of dates, numbers, and strings.
    """
    import pandas as pd

    # Handle datetime
    if isinstance(value, (pd.Timestamp, datetime)):
        # ISO 8601 format for datetime columns
        if column_name.endswith("DTC"):
            return value.isoformat() if hasattr(value, "isoformat") else str(value)
        return str(value)

    # Handle numeric values
    if isinstance(value, float):
        # Remove trailing zeros for cleaner output
        if value == int(value):
            return str(int(value))
        return f"{value:g}"

    if isinstance(value, int):
        return str(value)

    # Default string conversion
    return str(value)


def write_dataset_xml(
    data: pd.DataFrame,
    domain_code: str,
    config: MappingConfig,
    output: Path,
    *,
    metadata_version_oid: str | None = None,
    is_reference_data: bool | None = None,
) -> None:
    """Write a Dataset-XML 1.0 file for a single domain.

    Args:
        data: The pandas DataFrame containing the domain data.
        domain_code: The SDTM domain code (e.g., "DM", "AE").
        config: The mapping configuration containing study metadata.
        output: The output file path.
        metadata_version_oid: The MetaDataVersionOID to reference Define-XML.
        is_reference_data: Override for reference data detection.
    """
    domain = get_domain(domain_code)

    # Auto-detect reference data if not specified
    class_name = (domain.class_name or "").replace("-", " ").strip().upper()
    if is_reference_data is None:
        is_reference_data = class_name in ("TRIAL DESIGN", "STUDY REFERENCE")

    root = build_dataset_xml_tree(
        data,
        domain_code,
        config,
        metadata_version_oid=metadata_version_oid,
        is_reference_data=is_reference_data,
    )

    tree = ET.ElementTree(root)
    output.parent.mkdir(parents=True, exist_ok=True)
    tree.write(output, xml_declaration=True, encoding="utf-8")


def generate_dataset_xml_streaming(
    data: pd.DataFrame,
    domain_code: str,
    config: MappingConfig,
    output: Path,
    *,
    chunk_size: int = 1000,
    metadata_version_oid: str | None = None,
    is_reference_data: bool | None = None,
) -> None:
    """Generate Dataset-XML with streaming for large datasets.

    This method writes the XML incrementally to handle large datasets
    without loading the entire document into memory.

    Args:
        data: The pandas DataFrame containing the domain data.
        domain_code: The SDTM domain code.
        config: The mapping configuration.
        output: The output file path.
        chunk_size: Number of rows to process at a time.
        metadata_version_oid: The MetaDataVersionOID to reference Define-XML.
        is_reference_data: Override for reference data detection.
    """
    domain = get_domain(domain_code)
    study_id = (config.study_id or "STUDY").strip() or "STUDY"
    study_oid = f"STDY.{study_id}"
    dataset_name = domain.resolved_dataset_name()
    timestamp = datetime.now(UTC).isoformat(timespec="seconds")
    mdv_oid = metadata_version_oid or f"MDV.{study_oid}.SDTMIG.{DEFAULT_SDTM_VERSION}"
    define_file_oid = f"{study_oid}.Define-XML_{DEFINE_XML_VERSION}"

    # Auto-detect reference data
    class_name = (domain.class_name or "").replace("-", " ").strip().upper()
    if is_reference_data is None:
        is_reference_data = class_name in ("TRIAL DESIGN", "STUDY REFERENCE")

    output.parent.mkdir(parents=True, exist_ok=True)

    with output.open("w", encoding="utf-8") as f:
        # Write XML declaration
        f.write('<?xml version="1.0" encoding="UTF-8"?>\n')

        # Write ODM opening tag with namespaces
        f.write(f'<ODM xmlns="{ODM_NS}" ')
        f.write(f'xmlns:xlink="{XLINK_NS}" ')
        f.write(f'xmlns:data="{DATA_NS}" ')
        f.write('FileType="Snapshot" ')
        f.write(f'FileOID="{define_file_oid}(IG.{dataset_name})" ')
        f.write(f'PriorFileOID="{define_file_oid}" ')
        f.write('Granularity="All" ')
        f.write('ODMVersion="1.3.2" ')
        f.write(f'CreationDateTime="{timestamp}" ')
        f.write('Originator="CDISC-Transpiler" ')
        f.write('SourceSystem="CDISC-Transpiler" ')
        f.write('SourceSystemVersion="1.0" ')
        f.write(f'data:DatasetXMLVersion="{DATASET_XML_VERSION}">\n')

        # Write data container opening
        container_tag = "ReferenceData" if is_reference_data else "ClinicalData"
        f.write(f'  <{container_tag} StudyOID="{study_oid}" ')
        f.write(f'MetaDataVersionOID="{mdv_oid}">\n')

        # Stream data in chunks
        seq = 0

        for chunk_start in range(0, len(data), chunk_size):
            chunk = data.iloc[chunk_start : chunk_start + chunk_size]

            for _, row in chunk.iterrows():
                seq += 1
                _write_item_group_data_streaming(f, row, domain, dataset_name, seq)

        # Close tags
        f.write(f"  </{container_tag}>\n")
        f.write("</ODM>\n")


def _write_item_group_data_streaming(
    f: object,
    row: object,
    domain: SDTMDomain,
    dataset_name: str,
    seq: int,
) -> None:
    """Write a single ItemGroupData element to the file stream."""
    f.write(f'    <ItemGroupData ItemGroupOID="IG.{dataset_name}" ')
    f.write(f'data:ItemGroupDataSeq="{seq}">\n')

    for col_name, value in row.items():
        if _is_null(value):
            continue

        formatted_value = _format_value(value, col_name)
        # Escape XML special characters
        escaped_value = _escape_xml(formatted_value)

        item_oid = _generate_item_oid(col_name, domain.code)
        f.write(f'      <ItemData ItemOID="{item_oid}" ')
        f.write(f'Value="{escaped_value}"/>\n')

    f.write("    </ItemGroupData>\n")


def _escape_xml(value: str) -> str:
    """Escape special XML characters in attribute values."""
    return (
        value.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&apos;")
    )


def iter_dataset_xml_records(
    data: pd.DataFrame,
    domain_code: str,
) -> Iterator[ET.Element]:
    """Yield ItemGroupData elements for each record.

    This generator can be used to build custom Dataset-XML documents
    or to stream records to other formats.

    Args:
        data: The pandas DataFrame containing the domain data.
        domain_code: The SDTM domain code.

    Yields:
        ItemGroupData Element for each row in the DataFrame.
    """
    domain = get_domain(domain_code)
    dataset_name = domain.resolved_dataset_name()

    for seq, (_, row) in enumerate(data.iterrows(), start=1):
        item_group_data = ET.Element(
            _tag(ODM_NS, "ItemGroupData"),
            attrib={"ItemGroupOID": f"IG.{dataset_name}"},
        )
        item_group_data.set(_attr(DATA_NS, "ItemGroupDataSeq"), str(seq))

        for col_name in data.columns:
            value = row[col_name]
            if _is_null(value):
                continue

            formatted_value = _format_value(value, col_name)
            ET.SubElement(
                item_group_data,
                _tag(ODM_NS, "ItemData"),
                attrib={
                    "ItemOID": f"IT.{domain.code}.{col_name}",
                    "Value": formatted_value,
                },
            )

        yield item_group_data

File: crates/sdtm-output/src/dataset_xml.rs
//! Dataset-XML output generation.

use std::fs::File;
use std::io::BufWriter;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, anyhow};
use chrono::{SecondsFormat, Utc};
use polars::prelude::AnyValue;
use quick_xml::Writer;
use quick_xml::events::{BytesDecl, BytesEnd, BytesStart, Event};

use sdtm_ingest::any_to_string_non_empty;
use sdtm_model::Domain;
use sdtm_transform::{DomainFrame, domain_map_by_code};

use crate::common::{
    DATASET_XML_NS, DATASET_XML_VERSION, DEFINE_XML_VERSION, ODM_NS, XLINK_NS, ensure_output_dir,
    ensure_parent_dir, is_reference_domain, normalize_study_id,
};

/// Options for Dataset-XML output.
#[derive(Debug, Clone, Default)]
pub struct DatasetXmlOptions {
    pub dataset_name: Option<String>,
    pub metadata_version_oid: Option<String>,
    pub is_reference_data: Option<bool>,
}

/// Write Dataset-XML outputs for all domains.
pub fn write_dataset_xml_outputs(
    output_dir: &Path,
    domains: &[Domain],
    frames: &[DomainFrame],
    study_id: &str,
    sdtm_ig_version: &str,
) -> Result<Vec<PathBuf>> {
    let domain_lookup = domain_map_by_code(domains);
    let mut frames_sorted: Vec<&DomainFrame> = frames.iter().collect();
    frames_sorted.sort_by(|a, b| a.domain_code.cmp(&b.domain_code));

    let xml_dir = ensure_output_dir(output_dir, "dataset-xml")?;

    let mut outputs = Vec::new();
    for frame in frames_sorted {
        let code = frame.domain_code.to_uppercase();
        let domain = domain_lookup
            .get(&code)
            .ok_or_else(|| anyhow!("missing domain definition for {code}"))?;
        // Use frame's dataset name (from metadata) for split domains
        let output_dataset_name = frame.dataset_name();
        let disk_name = output_dataset_name.to_lowercase();
        let path = xml_dir.join(format!("{disk_name}.xml"));
        let options = DatasetXmlOptions {
            dataset_name: Some(output_dataset_name),
            ..Default::default()
        };
        write_dataset_xml(
            &path,
            domain,
            frame,
            study_id,
            sdtm_ig_version,
            Some(&options),
        )?;
        outputs.push(path);
    }
    Ok(outputs)
}

/// Write a single Dataset-XML file.
pub fn write_dataset_xml(
    output_path: &Path,
    domain: &Domain,
    frame: &DomainFrame,
    study_id: &str,
    sdtm_ig_version: &str,
    options: Option<&DatasetXmlOptions>,
) -> Result<()> {
    let options = options.cloned().unwrap_or_default();
    let dataset_name = options
        .dataset_name
        .unwrap_or_else(|| crate::common::dataset_name(domain));
    let study_id = normalize_study_id(study_id);
    let study_oid = format!("STDY.{study_id}");
    let mdv_oid = options
        .metadata_version_oid
        .unwrap_or_else(|| format!("MDV.{study_oid}.SDTMIG.{sdtm_ig_version}"));
    let define_file_oid = format!("{study_oid}.Define-XML_{DEFINE_XML_VERSION}");
    let file_oid = format!("{define_file_oid}(IG.{dataset_name})");
    let is_reference = options
        .is_reference_data
        .unwrap_or_else(|| is_reference_domain(domain));
    let container_name = if is_reference {
        "ReferenceData"
    } else {
        "ClinicalData"
    };

    let timestamp = Utc::now().to_rfc3339_opts(SecondsFormat::Secs, true);
    ensure_parent_dir(output_path)?;
    let file =
        File::create(output_path).with_context(|| format!("create {}", output_path.display()))?;
    let writer = BufWriter::new(file);
    let mut xml = Writer::new_with_indent(writer, b' ', 2);

    xml.write_event(Event::Decl(BytesDecl::new("1.0", Some("UTF-8"), None)))?;

    let mut root = BytesStart::new("ODM");
    root.push_attribute(("xmlns", ODM_NS));
    root.push_attribute(("xmlns:xlink", XLINK_NS));
    root.push_attribute(("xmlns:data", DATASET_XML_NS));
    root.push_attribute(("data:DatasetXMLVersion", DATASET_XML_VERSION));
    root.push_attribute(("FileType", "Snapshot"));
    root.push_attribute(("FileOID", file_oid.as_str()));
    root.push_attribute(("PriorFileOID", define_file_oid.as_str()));
    root.push_attribute(("ODMVersion", "1.3.2"));
    root.push_attribute(("CreationDateTime", timestamp.as_str()));
    root.push_attribute(("Originator", "CDISC-Transpiler"));
    xml.write_event(Event::Start(root))?;

    let mut container = BytesStart::new(container_name);
    container.push_attribute(("StudyOID", study_oid.as_str()));
    container.push_attribute(("MetaDataVersionOID", mdv_oid.as_str()));
    xml.write_event(Event::Start(container))?;

    let df = &frame.data;
    let mut columns = Vec::with_capacity(domain.variables.len());
    for variable in &domain.variables {
        let series = df
            .column(variable.name.as_str())
            .with_context(|| format!("missing column {}", variable.name))?;
        columns.push(series);
    }

    for row_idx in 0..df.height() {
        let mut group = BytesStart::new("ItemGroupData");
        let group_oid = format!("IG.{dataset_name}");
        let group_seq = format!("{}", row_idx + 1);
        group.push_attribute(("ItemGroupOID", group_oid.as_str()));
        group.push_attribute(("data:ItemGroupDataSeq", group_seq.as_str()));
        xml.write_event(Event::Start(group))?;
        for (variable, column) in domain.variables.iter().zip(columns.iter()) {
            let value = column.get(row_idx).unwrap_or(AnyValue::Null);
            if let Some(text) = any_to_string_non_empty(value) {
                let mut item = BytesStart::new("ItemData");
                let item_oid = format!("IT.{dataset_name}.{}", variable.name);
                item.push_attribute(("ItemOID", item_oid.as_str()));
                item.push_attribute(("Value", text.as_str()));
                xml.write_event(Event::Empty(item))?;
            }
        }
        xml.write_event(Event::End(BytesEnd::new("ItemGroupData")))?;
    }

    xml.write_event(Event::End(BytesEnd::new(container_name)))?;
    xml.write_event(Event::End(BytesEnd::new("ODM")))?;
    Ok(())
}
